apps/dashr-ocean-optics/app.R

```diff
@@ -0,0 +1,580 @@
+  input(id = controls[[i]]$props$children[[2]]$props$id,
+        property = ifelse(
+          length(controls[[i]]$props$children[[2]]$props$value) > 0,
+          "value",
+          "on"
+        ))
+})
+
+# state list control elements
+controlValuesState <- lapply(1:length(controls), function(i) {
+  state(id = controls[[i]]$props$children[[2]]$props$id,
+        property = ifelse(
+          length(controls[[i]]$props$children[[2]]$props$value) > 0,
+          "value",
+          "on"
+        ))
+})
+
+# generates intensities
+SampleSpectrum <- function(scale, knobIntensity, x) {
+  scale * (exp(-1 * ( (x - 500) / 5) ** 2) +
+             0.01 * runif(length(x))) + knobIntensity * 10
+}
+
+
+############################
+# Callbacks
+############################
+
+# disable/enable the update button depending on whether options have changed
+app$callback(
+  output = list(id = "submit-button", property = "style"),
+  params = c(controlValues,
+             list(input(id = "submit-button", "n_clicks_timestamp"))),
+
+  function(...) {
+    args <- list(...)
+
+    argsTime <- args[[length(args)]]
+    now <- as.numeric(Sys.time()) * 1000
+
+    disabled <- list(
+      'color' = colors[['accent']],
+      'backgroundColor' = colors[['background']],
+      'cursor' = 'not-allowed'
+    )
+
+    enabled <- list(
+      'color' = colors[['background']],
+      'backgroundColor' = colors[['accent']],
+      'cursor' = 'pointer'
+    )
+
+    # if the button was recently clicked (less than a second ago), then
+    # it's safe to say that the callback was triggered by the button; so
+    # we have to "disable" it
+    if (now - argsTime < 500 && argsTime > 0) {
+      return(disabled)
+    } else {
+      return(enabled)
+    }
+  }
+)
+
+
+# disable/enable controls
+app$callback(
+  output = list(id = "controls", property = "children"),
+  params = list(input(id = "power-button", property = "on")),
+
+  function(pwr_on) {
+    lapply(1:length(controls),
+           function(i) {
+             controls[[i]]$props$children[[2]]$props$disabled <- !pwr_on
+             controls[[i]]
+           })
+  }
+)
+
+
+# disable/enable intensity knob
+app$callback(
+  output = list(id = "light-intensity-knob", property = "disabled"),
+  params = list(
+    input(id = "power-button", property = "on"),
+    input(id = "light-source-input", property = "value")
+  ),
+
+  function(pwr, ls) {
+    return(!(pwr && ls != "NULL" && ls != ""))
+  }
+)
+
+
+# send user-selected options to spectrometer
+app$callback(
+  output = list(id = "submit-status", property = "children"),
+  params = c(list(input(id = "submit-button", "n_clicks")),
+             controlValuesState,
+             list(state(id = "power-button", "on"))),
+
+  function(...) {
+    args <- list(...)
+
+    # handle `NULL` args`
+    nullCheck <- sapply(args, class)
+    nullCheck == "list"
+    if (sum(nullCheck == "list") > 0 ) {
+      args[[which(nullCheck == "list")]] <- "None"
```

Comment:
Here, I'd expect that we want to assign `NULL` instead of `"None"`, as the latter is a Python sort of representation for null objects.

The line `nullCheck == "list"` looks inert; I think this is just comparing whether `nullCheck` is identical to `"list"`, but nothing is done with this information. Probably safe to :hocho:.

For `sum(nullCheck == "list") > 0`, I might actually rewrite this as `if ("list" %in% nullCheck)`. The reason for this is that the latter will work regardless of the dimensionality of `nullCheck`:

```
> nullCheck <- c("list", "character", "numeric")
> "list" %in% nullCheck
[1] TRUE
> nullCheck <- NA
> "list" %in% nullCheck
[1] FALSE
> sum(nullCheck)
[1] NA
> sum(nullCheck) > 0
[1] NA
```

Reply:
Assigning directly `NULL` to `args[[which(nullCheck == "list")]] <- NULL` results in corresponding `args` index to be removed from the list. Since it causes in the following `for` loop assignments to shift, I replaced `"None"` to `"NULL"` to follow R representation.

https://github.com/plotly/dash-sample-apps/blob/c04df66e9187b950f4dd850932b88580cd2aa8c8/apps/dashr-ocean-optics/app.R#L419-L435